### PR TITLE
feat(tests): add code coverage reporting

### PR DESCRIPTION
**Pull request by guidojw**
### Summary

Adds code coverage reporting with [Codecov](https://about.codecov.io/).
Since coverage was already very high, I modified and added some tests to get it to 100%.

### CODE REVIEW

File: app/resources/v1/application_resource.rb [hunk 1]
```diff
@@ -79,9 +77,11 @@ def current_user
     context.fetch(:user)
   end
 
+  # :nocov:
   def current_application
```
Comment (guidojw) [inline app/resources/v1/application_resource.rb:81]: This method seems unused

File: app/resources/v1/group_resource.rb [hunk 2]
```diff
@@ -20,10 +20,8 @@ def description_camofied
   has_many :permissions
 
   filter :active, apply: ->(records, _value, _options) { records.active }
-  filter :kind, apply: ->(records, value, _options) { records.where(kind: value[0]) }
-  filter :administrative, (lambda do |records, value, _options|
-    records.where(administrative: value[0])
-  end)
+  filter :kind
```
Comment (guidojw) [inline app/resources/v1/group_resource.rb:23]: This is the default implementation, so redundant

File: spec/requests/v1/books_controller/index_spec.rb [hunk 3]
```diff
@@ -39,7 +39,19 @@
...
 
-      it_behaves_like '200 OK'
+      describe 'with subtitle' do
+        it_behaves_like '200 OK'
+      end
+
+      describe 'without subtitle' do
+        let(:request) do
+          VCR.use_cassette('retrieve_book_by_isbn_no_subtitle') do
+            get "#{record_url}?isbn=8713638034201"
```
Comment (guidojw) [inline spec/requests/v1/books_controller/index_spec.rb:49]: @wilco375 isbn_lookup doesn't work, I get a 403 status which probably means the bol.com API token is expired. Can you rotate it and PR it into the `credentials.yml.enc`?

**Comment by guidojw** [comment]
Converted to draft until bol.com API token problem is resolved.

File: spec/requests/v1/books_controller/index_spec.rb (hunk 3 above)
Comment (guidojw) [reply spec/requests/v1/books_controller/index_spec.rb]: Resolved in https://github.com/csvalpha/amber-api/pull/402

File: app/resources/v1/application_resource.rb (hunk 1 above)
Comment (DrumsnChocolate) [reply app/resources/v1/application_resource.rb:81]: this is an endpoint, right? I suspect that's why it isn't covered by tests (or do we test endpoints too?) Did you check whether amber-ui depends on it?

File: README.md [hunk 4]
```diff
@@ -2,6 +2,7 @@ Alpha AMBER API
 ================
 [![Continuous Integration](https://github.com/csvalpha/amber-api/actions/workflows/continuous-integration.yml/badge.svg)](https://github.com/csvalpha/amber-api/actions/workflows/continuous-integration.yml)
 [![Continuous Delivery](https://github.com/csvalpha/amber-api/actions/workflows/continuous-delivery.yml/badge.svg)](https://github.com/csvalpha/amber-api/actions/workflows/continuous-delivery.yml)
+[![codecov](https://codecov.io/github/csvalpha/amber-api/graph/badge.svg?token=FC2UFYSOON)](https://codecov.io/github/csvalpha/amber-api)
```
Comment (DrumsnChocolate) [inline README.md:5]:
```suggestion
[![codecov](https://codecov.io/github/csvalpha/amber-api/graph/badge.svg)](https://codecov.io/github/csvalpha/amber-api)
```

File: .github/workflows/continuous-integration.yml [hunk 5]
```diff
@@ -118,5 +121,21 @@ jobs:
...
+          fail_ci_if_error: true
+          token: ${{ secrets.CODECOV_TOKEN }}
+
+      - name: Upload coverage report artifact
+        if: ${{ !cancelled() }}
+        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        with:
+          name: coverage
+          path: coverage/
+          if-no-files-found: error
```
Comment (DrumsnChocolate) [inline .github/workflows/continuous-integration.yml:141]: do these result in some kind of comment on PRs, too?

**Review by DrumsnChocolate — APPROVED** [review]
do we need to refresh the action hashes in the workflow?

**Pull request merged**